### PR TITLE
DBG: check licenses

### LINKED ISSUE #85: Add license check action
### 📝 Description of the feature

Add a new action ``check-licenses``. This action should also be incorporated as the last step of the already existing action ``build-library``.

### 💡 Steps for implementing the feature

The content of this ``check-licenses`` action should be:

1. Define accepted licenses. For instance, contains ``MIT``, contains ``BSD``, etc. See the accepted licenses in pyansys/pyansys [here](https://github.com/pyansys/pyansys/blob/main/.github/workflows/dependency-reviewer.yml).
2. Use ``pip-licenses`` to list all packages with its licenses 
    2.1. Ignore some trusted packages with different licenses that are not accepted. This can be done using the [ignore package option](https://github.com/raimon49/pip-licenses#option-ignore-packages)
    2.2. List the remaining packages' licenses, see [DPF action](https://github.com/pyansys/pydpf-actions/blob/v2.2/check-licenses/action.yml). ``pip-licenses --summary`` can be used to only list the licenses and the number of packages under each license.
3. Parse the table content and check that all licenses are among the accepted ones (step 1).

### 🔗 Useful links and references

- https://github.com/pyansys/pyansys/blob/main/.github/workflows/dependency-reviewer.yml
- https://github.com/pyansys/pydpf-actions/blob/v2.2/check-licenses/action.yml
- https://github.com/raimon49/pip-licenses#option-ignore-packages

### PR DESCRIPTION
**Pull request by GuillemBarroso**
Fix #85.

Debugging check licenses behavior. 

The ci_cd,yml has been modified to only execute this action for debugging purposes.

### CODE REVIEW

File: check-licenses/action.yml
Comment (jorgepiloto) [inline check-licenses/action.yml:19]: Try to execute any Python module using `python -m ...`. This ensures that you use the modules within your environment and not the binaries.

File: check-licenses/action.yml
Comment (jorgepiloto) [inline check-licenses/action.yml:26]: No need to echo a CLI tool that already outputs to stdout.

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml:19]: Got it!

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml:26]: Thats why I added both, with echo (in line 26) and without it (in line 27), just to see if any of them was printing in the logs

Comment (GuillemBarroso) [reply check-licenses/action.yml]: I've removed the unnecessary echos. However, it looks like it outputs to stdout an empty string:

image

I am able to print the ``accepted_licenses`` variable (strings to find in the ``pip_licenses --summary``) before and after removing the last two characters. But then I get 3 empty lines for the 3 prints that was supposed to show.

Comment (jorgepiloto) [reply check-licenses/action.yml]: The issue is that you have a clean Python environment without any package installed in the CI.

Comment (jorgepiloto) [reply check-licenses/action.yml]: It is working fine. Try to install numpy and matplotlib, so the Python environment gets cluttered with dependencies and packages.

**Comment by jorgepiloto** [comment]
To complete this action:

- [ ] Add a `python-version` as input.
- [ ] Add also the `skip-install` option, to avoid installing the project. This is just in case we want to reuse this action inside another action which already installs the project.

File: check-licenses/action.yml
Comment (RobPasMue) [inline check-licenses/action.yml:9]: What about ignored packages? Shouldn't we provide them as a list as well? The actions should also have a record file of accepted packages whose licenses are accepted (even if they are not of the types accepted by default). Don't know if you follow me but we can talk about this offline =)

Comment (GuillemBarroso) [reply check-licenses/action.yml:9]: yep, I'm on it. So far I'm only excluding matplotlib for testing purposes.

Now I'm working on making it fail if bad licenses are found. After that I will stop hard-coding the inputs and pass them properly. I am planning on adding these inputs:
- ``accepted-licenses``: with some default well-known licenses (I took the ones you had [here](https://github.com/pyansys/pyansys/blob/main/.github/workflows/dependency-reviewer.yml#L22), @RobPasMue).
- ``additional-accepted-licenses``: just in case some projects want to add one or two more to the ``accepted-licenses`` list.  Instead of having to re-write the entire list.
- ``ingored-packages``: list of packages that are trusted but have invalid licenses.
-  ``python-version``
- ``skip-install``: to avoid installing the package if this action is used inside another one that already installs the package

Then I will add the install package step before checking the licenses.

Do you think I'm missing something? Comments?

Comment (RobPasMue) [reply check-licenses/action.yml:9]: Well... It's confusing for me to have two arguments called ``accepted-licenses`` and ``additional-accepted-licenses``.

* The first ones (``accepted-licenses``) should be hard-coded in the action. In fact we should be reading them from a file (same way you execute a ``check-licenses.py`` file) so that it is more maintainable. We should keep track of them in a txt file or similar of the accepted licences by the PyAnsys repositories.
* The ``additional-accepted-licenses`` approach is good for those repos which have specific needs.
* The ``ignored-packages`` option should behave similarly to the previous two. We should have a list of generally accepted packages that have "wrong" licenses. And we should also allow users to provide additional packages (if needed)

Keep in mind that this action is oriented for PyAnsys repositories. There is no problem on tailoring it for our purpose - with these ``accepted-licenses.txt`` and ``ignored-packages.txt`` files, for example.

Comment (RobPasMue) [reply check-licenses/action.yml:9]: Don't know if you got my point but basically I am proposing to:
* Delete ``accepted-licenses`` and ``ignored-packages`` options
* Only accept ``additional-accepted-licenses`` and ``additional-ignored-packages``
* Write to a TXT file those licenses accepted by default, and read them from there in the action
* Write to a TXT file those packages accepted by default, and read them from there in the action

Comment (GuillemBarroso) [reply check-licenses/action.yml:9]: After an off-line conversation, we've decided to go for the txt approach.

Thus, the inputs of this actions are doing to be:
- ``python-version``
- ``skip-install``: to avoid installing the package if this action is used inside another one that already installs the package

Inside the action we're going to hard-code the most common licenses. Then, two txt files (``accepted-licenses.txt`` and ``ignored-packages.txt``) will be used in order to tweak which extra licenses are also accepted and which packages can be ignored, depending on the particularities of each repo.

Comment (GuillemBarroso) [reply check-licenses/action.yml:9]: pinging @jorgepiloto and @MaxJPRey for further input on the final decision.

Comment (RobPasMue) [reply check-licenses/action.yml:9]: My main point is that the default accepted licenses and ignored packages should already be defined inside this action (if possible, as TXT files). The way in which users provide additional licenses and ignored packages is up for discussion, either with an additional TXT file which they maintain in their own repos or directly in the YML file calling this action.

**Comment by GuillemBarroso** [comment]
The final approach that I've followed is to only accept licenses appearing in ``accepted-licenses.txt`` and ignore packages in ``ignored-packages.txt``. No additional licenses or packages can be added.

By doing this we force all pyansys repos to follow the same licenses. If a particular project wants to add a new license or ignored packages, they can do so by proposing a PR adding them in the txt files.

File: check-licenses/action.yml
Comment (MaxJPRey) [reply check-licenses/action.yml]: If you do so, just add a comment to explain why we need matplotlib and numpy.
@jorgepiloto Could numpy be enough?

Comment (GuillemBarroso) [reply check-licenses/action.yml]: Maybe I'm wrong, but this was just to test the action. Now I'm going to add all the libraries for PyAnsys2023.1.1 and I will remove the matplotlib and numpy imports. It shouldn't be required anymore when calling the action from other repos.

Comment (RobPasMue) [reply check-licenses/action.yml]: Yes, AFAIK, it was only for testing purposes.

**Comment by GuillemBarroso** [comment]
I have tested this action with the pyansys/pyansys repo, see [PR](https://github.com/pyansys/pyansys/pull/97).

They're both passing, so I believe we should be good to go. Please, let me know if you think something in the action could be improved.

File: .github/workflows/ci_cd.yml
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:24]: Remember to revert the main workflow.

File: accepted-licenses.txt
Comment (RobPasMue) [inline accepted-licenses.txt:1]: I think this file should just be deleted... right? It is duplicated. The only one that should live on is the one inside the ``check-licenses`` folder

File: ignored-packages.txt
Comment (RobPasMue) [inline ignored-packages.txt:2]: Same comment as for ``accepted-licenses.txt`` applies here.

File: check-licenses/action.yml
Comment (RobPasMue) [inline check-licenses/action.yml:20]: I left you a comment on the ``pyansys/pyansys`` PR. Please have a look at it: https://github.com/pyansys/pyansys/pull/97#issuecomment-1328592298

We might need to provide an additional optional argument called ``target``. Plan on implementing it.

**Review by RobPasMue — CHANGES_REQUESTED** [review]
I am missing documentation on how to use this action. Are you planning on adding it? I think it might be a good addition to your PR. Apart from my comments, LGTM @GuillemBarroso! Re-request review once the comments are implemented.

File: check-licenses/action.yml
Comment (jorgepiloto) [inline check-licenses/action.yml:50]: Note that these URLs need are currently pointing to the files in the branch you created, @GuillemBarroso. Once everything is ready to be merged, they should be changed assuming that the files are in `main`. This will make the actions to fail in this PR but pass once merged.

Comment (jorgepiloto) [reply check-licenses/action.yml:50]: Also, could we isolate the download of the files in a separate step? Just to split de setup process and the final logic of the licensing checking.

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: check-licenses/action.yml
Comment (RobPasMue) [reply check-licenses/action.yml:50]: Oh true! Yeah, those URLs have to be modified =)

And yeah, we might want to break down the entire step into more isolated ones. Good idea @jorgepiloto.

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: accepted-licenses.txt
Comment (GuillemBarroso) [reply accepted-licenses.txt:1]: Yes, that's right.

File: ignored-packages.txt
Comment (GuillemBarroso) [reply ignored-packages.txt:2]: Done.

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml]: Done.

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml:60]: @jorgepiloto, yes. This applies to both the links in ``action.yml`` and the @feat/chech-licenses in ``ci_cd.yml``.

What's the proper way of doing that? I was thinking on first merging the PR (should pass since the branch will still be "alive"), and then proposing another PR changing the two files mentioned above to point to @main with the correct links.

Finally, I think the links in ``action.yml`` have to be updated every time that the files are modified. Is this correct? (For example, now the PR in pyanys/pyansys is failing because it does not download the updated version of the files).

File: check-licenses/accepted-licenses.txt
Comment (GuillemBarroso) [reply check-licenses/accepted-licenses.txt]: Done.

File: doc/source/check-licenses-action/index.rst
Comment (jorgepiloto) [inline doc/source/check-licenses-action/index.rst:1]: Let us rename this to `Licenses actions` to be consistent with the rest of the headers. Rename the folder as well.

File: doc/source/index.rst
Comment (jorgepiloto) [inline doc/source/index.rst:10]: See previous comment about renaming this folder to unify the directories names.

File: .github/workflows/ci_cd.yml
Comment (jorgepiloto) [inline .github/workflows/ci_cd.yml:103]: We can not add this action within the CI pipeline of this repository. The reason is that this is not a Python package. Even if we use the `skip-install`, we are not creating any Python virtual environment or publishing third party packages.

This is the reason of the failing CI.

File: check-licenses/accepted-licenses.txt
Comment (jorgepiloto) [inline check-licenses/accepted-licenses.txt:10]:
```suggestion
```

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: doc/source/check-licenses-action/index.rst
Comment (RobPasMue) [inline doc/source/check-licenses-action/index.rst:2]:
```suggestion
================
```

File: doc/source/index.rst
Comment (GuillemBarroso) [reply doc/source/index.rst]: I have renamed the doc section to Licenses actions. However, I left the action itself as check-licenses. Let me know if you're OK with that @jorgepiloto.

File: doc/source/check-licenses-action/index.rst
Comment (RobPasMue) [reply doc/source/check-licenses-action/index.rst:2]: Remember to align title underline to title length

File: doc/source/index.rst
Comment (RobPasMue) [reply doc/source/index.rst]: Yeah that sounds fine, don't worry. One thing is the action itself, something different are the docs. The action name should be as explicit and standalone as possible

File: doc/source/licenses-actions/index.rst
Comment (jorgepiloto) [inline doc/source/licenses-actions/index.rst:11]: We are using one section per action. In addition, the title of the section is the title of the action, see the rest of the documentation files. Name this section to something like:

```suggestion
Check licenses action
```

All the content related with this action should be described below this section. For the guides on how to update the TXT files, take advantage of a `.. note::` directive.

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: check-licenses/action.yml
Comment (RobPasMue) [inline check-licenses/action.yml:59]: Are you sure that the moment we merge this PR (and the branch gets deleted), those links will still work? I'd say they won't... 
```suggestion
    - name: "Download txt files so they are found from other repos"
      shell: bash
      run: |
        wget https://raw.githubusercontent.com/pyansys/actions/main/check-licenses/check-licenses/accepted-licenses.txt
        wget https://raw.githubusercontent.com/pyansys/actions/main/check-licenses/ignored-packages.txt
```

You'll probably have to merge this PR with a failing status due to this breaking change. But once it's on main, it should work.

In fact, this link will have to be edited every time we do a release so that it points to whatever release branch we have... It ain't gonna be easy to maintain... We should think something... 

What about using this action to do that same operation? https://github.com/marketplace/actions/checkout-files

Comment (GuillemBarroso) [reply check-licenses/action.yml:59]: I have a question here that I was discussing with Maxime on Friday.

Won't it work if we leave this as feat/check-licenses for now? It will still point to the feat branch, but this branch will still be available until it gets removed right? I was thinking about first merging and then opening another PR to change from feat/check-licenses to main. Am I missing something?

I really like the approach you suggested with the checkout files action. I am going to implement it now.

Comment (RobPasMue) [reply check-licenses/action.yml:59]: Sure, we can do it in 2 PRs =) there is no issue in that. I was just trying to optimize. But at least we will get our green ticks dopamine 😄 

And yeah, let's try to use the action. I think it'll be more clear for everybody to understand what this step is doing. I know what you're doing, and probably anybody that understand what ``wget`` is knows what you're doing... but just in case =)

File: doc/source/licenses-actions/index.rst
Comment (GuillemBarroso) [reply doc/source/licenses-actions/index.rst]: Done.

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml:59]: For now, I'll leave it like this since I want to test the action from the PR in ``pyansys/pyansys``. After checking that everything works, I'll modify all the references to @feat/check-licenses to @main.

File: doc/source/licenses-actions/index.rst
Comment (RobPasMue) [inline doc/source/licenses-actions/index.rst:38]:
```suggestion
  that is not included in the original list, a PR can be proposed at ``pyansys/actions`` to
```

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml:69]: Nice, thanks for the suggestion.

Comment (GuillemBarroso) [reply check-licenses/action.yml:69]: hm.. wait. Beacuse ``licenses_from_txt`` is an array... it's not a string with spaces between licenses. Will this work in this case? Same for ``ignored_packages_from_txt``.

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml]: @RobPasMue. I have implemented the [action](https://github.com/marketplace/actions/checkout-files) following the documentation. However, when using the ``check-licenses`` action in the [pyansys/pyansys PR](https://github.com/pyansys/pyansys/pull/97) it looks like it is trying to find the files in the ``pyansys/pyansys`` repo, see url that is trying to access:

image

I don't see an option to define the repo in which the files are located... Is it possible that this action is only capable of checking out files that are in the same repo where the action is called?

Comment (jorgepiloto) [reply check-licenses/action.yml]: I think there shouldn't be a problem to define these files in the `main` branch, even if those point from a release branch. They are expected to be quite stable.

Comment (RobPasMue) [reply check-licenses/action.yml]: Ha true... it looks like this action is only valid for files in the same repo... sorry for that =( you might need to revert to the ``wget`` mechanism

Comment (RobPasMue) [reply check-licenses/action.yml]: Summarizing:
* Let's revert to the ``wget`` mechanism
* Point directly to its expected location in the ``main`` branch
* Merge whenever you feel it's ready =)

**Comment by RobPasMue** [comment]
Should we merge it then? Or are we waiting for something else? Let me do a final review just in case

File: doc/source/licenses-actions/index.rst
Comment (RobPasMue) [inline doc/source/licenses-actions/index.rst:16]:
```suggestion
This action accepts the following inputs.
```

File: doc/source/licenses-actions/index.rst
Comment (RobPasMue) [inline doc/source/licenses-actions/index.rst:37]:
```suggestion
  If a certain repository requires a different license or package
```

**Review by RobPasMue — APPROVED** [review]
Some comments, otherwise LGTM! Thanks @GuillemBarroso for all the iterations!

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: check-licenses/action.yml
Comment (GuillemBarroso) [reply check-licenses/action.yml]: No probs! Done.

**Comment by GuillemBarroso** [comment]
I've checked that the action works properly when calling it from ``pyansys/pyansys``. 

I will now change all references to feat/check-licenses to main and it'll be ready to merge.

**Review by RobPasMue — APPROVED** [review]
LGTM! Let's go!

**Review by jorgepiloto — APPROVED** [review]

**Pull request merged**
